### PR TITLE
feature: enabling training only the last N layers of the text encoder

### PR DESCRIPTION
**Pull request by damian0815**
Enabled by setting `last_n_layers` to an integer (eg `2` to train only the last 2 layers) in the `text_encoder_overrides` block of your `optimizer.json`.

Draft state because: 
* is this the right place for the `last_n_layers` param?
* breaks resume - `load_optimizer_state` bails with `ValueError: loaded state dict contains a parameter group that doesn’t match the size of optimizer’s group` which suggests i'm not setting it up exactly right (maybe need to explicitly freeze the other parameters?)

### CODE REVIEW

File: optimizer.json [hunk 1]
```diff
@@ -33,5 +37,11 @@
         "betas": null,
         "epsilon": null,
         "weight_decay": null
+    },
+    "text_encoder_freezing": {
+        "freeze_embeddings": false,
+        "freeze_front_n_layers": null,
+        "freeze_final_layer_norm": true,
+        "method": "drop"
```
Comment (victorchall) [inline optimizer.json:45]: Think you mentioned on discord, but if this ends up staying make sure to add some documentation for it.

Comment (damian0815) [reply optimizer.json:45]: yeah "method" will be gone as soon as i figure out which to use and how to reliably resume optimizer state with it

**Comment by damian0815** [comment]
@victorchall this is good to go. the parameters are a bit fiddly, i played around with how it felt if i reversed them (i.e. naming the arguments "unfreeze" instead of "freeze" and counting unfreezing layers backwards from the end), but it wasn't really much improvement.

i also took the liberty of enabling partial text encoder freezing by default in the `trainSD21.json`, and boosting the LR to match.

File: optimizer/optimizers.py [hunk 2]
```diff
@@ -298,12 +304,48 @@ def _create_optimizer(self, args, local_optimizer_config, parameters):
...
+        log_optimizer(label, optimizer, betas, epsilon, weight_decay, curr_lr)
         return optimizer
 
-def log_optimizer(optimizer: torch.optim.Optimizer, betas, epsilon, weight_decay, lr):
+    def _apply_text_encoder_freeze(self, text_encoder) -> chain[Any]:
+        parameters = itertools.chain([])
+
+        if self.te_freeze_config.get('freeze_embeddings', False):
+            # freeze embeddings
+            print(" ❄️ freezing embeddings")
```
Comment (victorchall) [inline optimizer/optimizers.py:315]: The emoji is cute, but will it cause problems on Colab or Vast?

Comment (damian0815) [reply optimizer/optimizers.py:315]: i checked Jupyter and colab and they both handle it fine

**Review by victorchall — APPROVED** [review]

**Pull request merged**